Assets/Scripts/Framework/Train/TrainEngine.cs

```diff
@@ -0,0 +1,236 @@
+        /// Current train engine state.
+        /// </summary>
+        public TrainEngineState EngineState 
+            { get; private set; } = TrainEngineState.OnStandby;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train speed type.
+        /// </summary>
+        public TrainEngineSpeedTypes CurrentTargetEngineSpeedType 
+            { get; private set; } = TrainEngineSpeedTypes.Still;
+
+        /// <summary>
+        /// Current train velocity speed.
+        /// </summary>
+        public float CurrentVelocity => _currentSpeed / _trainController.SplineLength;
+
+        /// <summary>
+        /// Determines the chosen track for the next possible rail split.<br></br>
+        /// <br>False = left.</br>
+        /// <br>True = right.</br>
+        /// </summary>
+        public bool CurrentSplitDirection { get; set; }
+
+        /// <summary>
+        /// Invokes when train direction is changed.
+        /// </summary>
+        public Action<bool> OnDirectionChanged;
+
+        /// <summary>
+        /// Invoked when the train speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train target speed is changed.
+        /// </summary>
+        public Action<TrainEngineSpeedTypes> OnTargetSpeedChanged;
+
+        /// <summary>
+        /// Invoked when the train engine state is changed.
+        /// </summary>
+        public Action<TrainEngineState> OnEngineStateChanged;
+
+        private Dictionary<TrainEngineSpeedTypes, float> _speedValues;
+
+        [HideInInspector]
+        [SyncVar(Channel = FishNet.Transporting.Channel.Reliable)]
+        private float _currentSpeed;
+        
+        private float _speedTypesCount;
+
+        [SerializeField] private bool isLerping;
+
+        private Tween _speedTween;
+        private TrainController _trainController;
+
+        private void Awake() => _trainController = GetComponent<TrainController>();
+
+        private void Start()
+        {
+            _speedTypesCount = Enum.GetValues(typeof(TrainEngineSpeedTypes)).Length - 1;
+
+            // Initialize speed values dictionary.
+            _speedValues = new Dictionary<TrainEngineSpeedTypes, float>()
+            {
+                {TrainEngineSpeedTypes.High, _maxHighSpeed },
+                {TrainEngineSpeedTypes.Low, _maxLowSpeed },
+                {TrainEngineSpeedTypes.Medium, _maxMediumSpeed },
+                
+                {TrainEngineSpeedTypes.Still, 0 },
+
+                {TrainEngineSpeedTypes.LowReverse, -_maxLowSpeed },
+                {TrainEngineSpeedTypes.MediumReverse, -_maxMediumSpeed },
+                {TrainEngineSpeedTypes.HighReverse, -_maxHighSpeed },
+            };
```

Comment:
Can't you just write a return function using a switch case instead of using a dictionairy? Would be more appropriate I think.

Reply:
Personally I prefer a dictionary since I think it is a cleaner solution and performance is not an issue in this situation.

Second opinion @lucasvdiepen ?

Reply:
I prefer a dictionary aswell. I think that is cleaner in this situation.